src/components/AutoUpdateTime.js

```diff
@@ -22,85 +27,70 @@ const propTypes = {
-        this.updateCurrentTime();
-    }
-
-    componentDidUpdate() {
-        // Make sure the interval is up to date every time the component updates
-        this.updateCurrentTime();
-    }
-
-    componentWillUnmount() {
-        clearTimeout(this.timer);
-    }
-
+function AutoUpdateTime(props) {
     /**
      * @returns {moment} Returns the locale moment object
      */
-    getCurrentUserLocalTime() {
-        return DateUtils.getLocalMomentFromDatetime(
-            this.props.preferredLocale,
+    const getCurrentUserLocalTime = useCallback(() => (
+        DateUtils.getLocalMomentFromDatetime(
+            props.preferredLocale,
             null,
-            this.props.timezone.selected,
-        );
-    }
+            props.timezone.selected,
+        )
+    ), [props.preferredLocale, props.timezone.selected]);
+
+    const [currentUserLocalTime, setCurrentUserLocalTime] = useState(getCurrentUserLocalTime());
+    const timerRef = useRef(null);
 
     /**
      * @returns {string} Returns the timezone name in string, e.g.: GMT +07
      */
-    getTimezoneName() {
+    const getTimezoneName = useCallback(() => {
         // With non-GMT timezone, moment.zoneAbbr() will return the name of that timezone, so we can use it directly.
-        if (Number.isNaN(Number(this.state.currentUserLocalTime.zoneAbbr()))) {
-            return this.state.currentUserLocalTime.zoneAbbr();
+        if (Number.isNaN(Number(currentUserLocalTime.zoneAbbr()))) {
+            return currentUserLocalTime.zoneAbbr();
         }
 
         // With GMT timezone, moment.zoneAbbr() will return a number, so we need to display it as GMT {abbreviations} format, e.g.: GMT +07
-        return `GMT ${this.state.currentUserLocalTime.zoneAbbr()}`;
-    }
+        return `GMT ${currentUserLocalTime.zoneAbbr()}`;
+    }, [currentUserLocalTime]);
 
     /**
      * Update the user's local time at the top of every minute
      */
-    updateCurrentTime() {
-        if (this.timer) {
-            clearTimeout(this.timer);
-            this.timer = null;
+    const updateCurrentTime = useCallback(() => {
+        if (timerRef.current) {
+            clearTimeout(timerRef.current);
+            timerRef.current = null;
         }
-        const millisecondsUntilNextMinute = (60 - this.state.currentUserLocalTime.seconds()) * 1000;
-        this.timer = setTimeout(() => {
-            this.setState({
-                currentUserLocalTime: this.getCurrentUserLocalTime(),
-            });
+        const millisecondsUntilNextMinute = (60 - currentUserLocalTime.seconds()) * 1000;
+        timerRef.current = setTimeout(() => {
+            setCurrentUserLocalTime(getCurrentUserLocalTime());
         }, millisecondsUntilNextMinute);
-    }
+    }, [currentUserLocalTime, getCurrentUserLocalTime]);
+
+    useEffect(() => {
+        updateCurrentTime();
+
+        return () => {
+            clearTimeout(timerRef.current);
+        };
+    }, [updateCurrentTime]);
```

Comment:
If we want this code to be equivalent to what it was before we should maybe remove all the dependencies. Since `updateCurrentTime()` after every update before. Now it only runs when the reference for `updateCurrentTime` changes (i.e. `useCallback()` returns a new function because the `currentUserLocalTime` changed).

Putting it all together that would end up just looking like this:

```js
useEffect(() => {
    if (timerRef.current) {
        clearTimeout(timerRef.current);
        timerRef.current = null;
    }
    const millisecondsUntilNextMinute = (60 - currentUserLocalTime.seconds()) * 1000;
    timerRef.current = setTimeout(() => {
        setCurrentUserLocalTime(getCurrentUserLocalTime());
    }, millisecondsUntilNextMinute);

    return () => {
        clearTimeout(timerRef.current);
    };
});
```

Reply:
That said, what you have here might work the same - but I didn't think too much about it.

Reply:
I agree that this may be an equivalent of the previous class implementation, but I believe while creating functional component we should slightly change the mindset and instead of mapping class component's functionality into functional one we should try to re-write the same functionality from scratch using new tools. 

Previously the above code was called on `componentDidMount` and `componentDidUpdate` lifecycle events - we should forget about lifecycle (not completely, just keep it in mind it exists and may have an impact on our code) and think when (on what prop/state change) this effect should be called.

So in the proposed code, I would add the dependency array - maybe empty one will be enough, maybe it should depend on some props changes (excluding `currentUserLocalTime` - please see my other comment). But leaving it without the deps array at all will cause it to be called on every single re-render which should be avoided.

Reply:
I think it's kind of strange actually now that I've looked at it and taking into consideration @burczu's notes. Maybe we should use a `setInterval()` instead to run the logic once per minute?

Reply:
But I guess we want it to run at the "top of every minute" instead of ~every minute...

Reply:
Might be interesting to have a custom hook that computes the value at the correct interval and returns it

```js
function useOncePerMinute(callback, dependencies) {
    const [computed, setComputed] = useState(callback);
    const timerRef = useRef(null);
    useEffect(() => {
        timerRef.current = setInterval(() => setComputed(callback()), 60000);
        return () => clearInterval(timerRef.current);
    }, dependencies);
    return computed;
}

const currentUserLocalTime = useOncePerMinute(getCurrentUserLocalTime, [props.preferredLocale, props.timezone.selected]);
```

Reply:
Yeah, that could work. From what I'm reading, intervals are not _super_ precise. Another option would be to set an interval that checks every second whether the minute has changed, and if so, update the state. Like this: 

```js
    const minuteRef = useRef(new Date().getMinutes());
    useEffect(() => {
        const interval = setInterval(() => {
            const now = new Date().getMinutes();
            if (now > minuteRef.current) {
                setCurrentUserLocalTime(getCurrentUserLocalTime());
            }
        }, 1000);
        return () => clearInterval(interval);
    }, [getCurrentUserLocalTime]);
```

Reply:
eh sorry I don't really get it tbh. `minuteRef` never gets updated and `now` will always be greater so we're just updating once per second - unless I'm missing something.

Reply:
Though if we care about precision maybe something like this could help https://justinmahar.github.io/react-use-precision-timer/?path=/story/home--page

Reply:
Lol whoops. Added a line for: 

```minuteRef.current = now;```

I think all work good enough, including what is there currently. It's just a matter of which we like best.

Reply:
We can see what @burczu thinks as well.

Reply:
Ok haha I see it now. Looks like it can work. Yeah I don't really have super strong feelings but think some kind of interval makes more sense than a timeout.

Reply:
Ach, you guys are right - using `setInterval` makes much more sense in this scenario! And i like the idea of extracting it into separate hook.

Reply:
I thought about it more and while I think the separate hook works, I think I prefer the other method for two reasons: 
1. It updates the clock at the top of the minute, rather than every 60 seconds from when the component is loaded 
2. It's the most simple/straightforward

If y'all strongly disagree with that, let me know. Otherwise, this is ready for re-review!

Reply:
tbh I'm fine with whatever. we kinda over-engineered it because the component is _only_ used on the Details page. It's hard to imagine anyone is going to sit there for minutes at a time or that anyone will notice that the time is updated at the precise top of the minute.